diff -Napur linux-6.6.6-orig/drivers/net/usb/lan78xx.c linux-6.6.6/drivers/net/usb/lan78xx.c
--- linux-6.6.6-orig/drivers/net/usb/lan78xx.c	2023-12-11 01:40:17.000000000 -0800
+++ linux-6.6.6/drivers/net/usb/lan78xx.c	2023-12-11 10:55:17.804354566 -0800
@@ -414,6 +414,28 @@ struct irq_domain_data {
 	struct mutex		irq_lock;		/* for irq bus access */
 };
 
+#ifdef CONFIG_KSZ_SWITCH
+
+/* It is expected that IBA_KSZ9897 is chosen for this MAC driver. */
+#if 0
+#ifndef CONFIG_HAVE_KSZ9897
+#define CONFIG_HAVE_KSZ9897
+#endif
+#ifndef CONFIG_KSZ_IBA
+#define CONFIG_KSZ_IBA
+#endif
+#ifndef CONFIG_IBA_KSZ9897
+#define CONFIG_IBA_KSZ9897
+#endif
+#ifndef CONFIG_KSZ_SWITCH_EMBEDDED
+#define CONFIG_KSZ_SWITCH_EMBEDDED
+#endif
+#endif
+
+#include <linux/platform_device.h>
+#include "../ethernet/micrel/ksz_mac_pre.c"
+#endif
+
 struct lan78xx_net {
 	struct net_device	*net;
 	struct usb_device	*udev;
@@ -475,8 +497,72 @@ struct lan78xx_net {
 	struct statstage	stats;
 
 	struct irq_domain_data	domain_data;
+
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_mac		sw_mac;
+#endif
 };
 
+#ifdef CONFIG_KSZ_SWITCH
+static struct ksz_mac *get_ksz_mac(void *ptr)
+{
+	struct lan78xx_net *dev = ptr;
+
+	return &dev->sw_mac;
+}
+
+static struct ksz_mac *get_netdev_priv(struct net_device *net)
+{
+	return get_ksz_mac(netdev_priv(net));
+}
+
+static void sw_reset_mac_mib(struct ksz_mac *priv)
+{
+	struct lan78xx_net *dev = priv->dev;
+
+	/* Need command to really clear hardware MIB counters. */
+	mutex_lock(&dev->stats.access_lock);
+	memset(&dev->stats.curr_stat, 0, sizeof(dev->stats.curr_stat));
+	mutex_unlock(&dev->stats.access_lock);
+}
+
+static void setup_ksz_mac(struct lan78xx_net *dev, struct net_device *net)
+{
+	dev->sw_mac.dev = dev;
+	dev->sw_mac.net = net;
+	dev->sw_mac.hw_priv = &dev->sw_mac;
+}
+
+static inline struct ksz_sw *get_sw(struct lan78xx_net *dev)
+{
+	return dev->sw_mac.port.sw;
+}
+
+static inline struct lan78xx_net *get_hw_dev(struct lan78xx_net *dev)
+{
+	return dev->sw_mac.hw_priv->dev;
+}
+
+static inline bool is_virt_dev(struct lan78xx_net *dev)
+{
+	struct ksz_mac *priv = get_ksz_mac(dev);
+
+	if (priv->hw_priv && priv != priv->hw_priv)
+		return true;
+	return false;
+}
+
+#ifndef CONFIG_KSZ_SWITCH_EMBEDDED
+#define USE_SPEED_LINK
+#define USE_MIB
+#endif
+
+#define KSZ_USE_MSGLEVEL
+#define KSZ_USE_IOCTL
+#define KSZ_USE_PRIVATE_IOCTL
+#include "../ethernet/micrel/ksz_mac.c"
+#endif
+
 /* define external phy id */
 #define	PHY_LAN8835			(0x0007C130)
 #define	PHY_KSZ9031RNX			(0x00221620)
@@ -1268,6 +1354,39 @@ static void lan78xx_deferred_multicast_w
 	lan78xx_write_reg(dev, RFE_CTL, pdata->rfe_ctl);
 }
 
+#ifdef CONFIG_KSZ_SWITCH
+static void hw_set_multicast(void *hw_priv, int multicast)
+{
+	struct lan78xx_net *dev = hw_priv;
+	struct lan78xx_priv *pdata = (struct lan78xx_priv *)(dev->data[0]);
+
+	if (multicast) {
+		pdata->rfe_ctl |= RFE_CTL_MCAST_EN_;
+	} else {
+		if (!dev->sw_mac.hw_promisc)
+			pdata->rfe_ctl &= ~RFE_CTL_MCAST_EN_;
+	}
+
+	schedule_work(&pdata->set_multicast);
+}  /* hw_set_multicast */
+
+static void hw_set_promisc(void *hw_priv, int promisc)
+{
+	struct lan78xx_net *dev = hw_priv;
+	struct lan78xx_priv *pdata = (struct lan78xx_priv *)(dev->data[0]);
+
+	if (promisc) {
+		pdata->rfe_ctl |= RFE_CTL_MCAST_EN_ | RFE_CTL_UCAST_EN_;
+	} else {
+		pdata->rfe_ctl &= ~(RFE_CTL_MCAST_EN_ | RFE_CTL_UCAST_EN_);
+		if (dev->sw_mac.hw_multi)
+			pdata->rfe_ctl |= RFE_CTL_MCAST_EN_;
+	}
+
+	schedule_work(&pdata->set_multicast);
+}  /* hw_set_promisc */
+#endif
+
 static void lan78xx_set_multicast(struct net_device *netdev)
 {
 	struct lan78xx_net *dev = netdev_priv(netdev);
@@ -1275,6 +1394,11 @@ static void lan78xx_set_multicast(struct
 	unsigned long flags;
 	int i;
 
+#ifdef CONFIG_KSZ_SWITCH
+	if (sw_set_rx_mode(netdev))
+		return;
+#endif
+
 	spin_lock_irqsave(&pdata->rfe_ctl_lock, flags);
 
 	pdata->rfe_ctl &= ~(RFE_CTL_UCAST_EN_ | RFE_CTL_MCAST_EN_ |
@@ -1554,6 +1678,10 @@ static int lan78xx_ethtool_get_eeprom(st
 	struct lan78xx_net *dev = netdev_priv(netdev);
 	int ret;
 
+#ifdef CONFIG_KSZ_SWITCH
+	dev = get_hw_dev(dev);
+#endif
+
 	ret = usb_autopm_get_interface(dev->intf);
 	if (ret)
 		return ret;
@@ -1573,6 +1701,10 @@ static int lan78xx_ethtool_set_eeprom(st
 	struct lan78xx_net *dev = netdev_priv(netdev);
 	int ret;
 
+#ifdef CONFIG_KSZ_SWITCH
+	dev = get_hw_dev(dev);
+#endif
+
 	ret = usb_autopm_get_interface(dev->intf);
 	if (ret)
 		return ret;
@@ -1613,6 +1745,10 @@ static void lan78xx_get_stats(struct net
 {
 	struct lan78xx_net *dev = netdev_priv(netdev);
 
+#ifdef CONFIG_KSZ_SWITCH
+	dev = get_hw_dev(dev);
+#endif
+
 	lan78xx_update_stats(dev);
 
 	mutex_lock(&dev->stats.access_lock);
@@ -1628,6 +1764,17 @@ static void lan78xx_get_wol(struct net_d
 	u32 buf;
 	struct lan78xx_priv *pdata = (struct lan78xx_priv *)(dev->data[0]);
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_sw *sw = get_sw(dev);
+
+	/* Not implemented yet. */
+	if (sw_is_switch(sw)) {
+		wol->supported = 0;
+		wol->wolopts = 0;
+		return;
+	}
+#endif
+
 	if (usb_autopm_get_interface(dev->intf) < 0)
 		return;
 
@@ -1655,6 +1802,15 @@ static int lan78xx_set_wol(struct net_de
 	struct lan78xx_priv *pdata = (struct lan78xx_priv *)(dev->data[0]);
 	int ret;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_sw *sw = get_sw(dev);
+
+	/* Not implemented yet. */
+	if (sw_is_switch(sw)) {
+		return 0;
+	}
+#endif
+
 	ret = usb_autopm_get_interface(dev->intf);
 	if (ret < 0)
 		return ret;
@@ -1680,6 +1836,18 @@ static int lan78xx_get_eee(struct net_de
 	int ret;
 	u32 buf;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_sw *sw = get_sw(dev);
+
+	if (sw_is_switch(sw)) {
+		edata->eee_enabled = false;
+		edata->eee_active = false;
+		edata->tx_lpi_enabled = false;
+		edata->tx_lpi_timer = 0;
+		return 0;
+	}
+#endif
+
 	ret = usb_autopm_get_interface(dev->intf);
 	if (ret < 0)
 		return ret;
@@ -1717,6 +1885,14 @@ static int lan78xx_set_eee(struct net_de
 	int ret;
 	u32 buf;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_sw *sw = get_sw(dev);
+
+	if (sw_is_switch(sw)) {
+		return 0;
+	}
+#endif
+
 	ret = usb_autopm_get_interface(dev->intf);
 	if (ret < 0)
 		return ret;
@@ -1745,6 +1921,14 @@ static u32 lan78xx_get_link(struct net_d
 {
 	u32 link;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct lan78xx_net *dev = netdev_priv(net);
+	struct ksz_sw *sw = get_sw(dev);
+
+	if (sw_is_switch(sw)) {
+		return netif_carrier_ok(net);
+	}
+#endif
 	mutex_lock(&net->phydev->lock);
 	phy_read_status(net->phydev);
 	link = net->phydev->link;
@@ -1766,6 +1950,13 @@ static u32 lan78xx_get_msglevel(struct n
 {
 	struct lan78xx_net *dev = netdev_priv(net);
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_sw *sw = get_sw(dev);
+
+	if (sw_is_switch(sw)) {
+		dev->msg_enable = dev->sw_mac.msg_enable;
+	}
+#endif
 	return dev->msg_enable;
 }
 
@@ -1773,6 +1964,9 @@ static void lan78xx_set_msglevel(struct
 {
 	struct lan78xx_net *dev = netdev_priv(net);
 
+#ifdef CONFIG_KSZ_SWITCH
+	sw_set_msglevel(net, &dev->sw_mac, level);
+#endif
 	dev->msg_enable = level;
 }
 
@@ -1783,6 +1977,15 @@ static int lan78xx_get_link_ksettings(st
 	struct phy_device *phydev = net->phydev;
 	int ret;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_sw *sw = get_sw(dev);
+
+	if (sw_is_switch(sw)) {
+		phy_ethtool_ksettings_get(phydev, cmd);
+		return 0;
+	}
+#endif
+
 	ret = usb_autopm_get_interface(dev->intf);
 	if (ret < 0)
 		return ret;
@@ -1802,6 +2005,15 @@ static int lan78xx_set_link_ksettings(st
 	int ret = 0;
 	int temp;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_sw *sw = get_sw(dev);
+
+	if (sw_is_switch(sw)) {
+		ret = phy_ethtool_ksettings_set(phydev, cmd);
+		return ret;
+	}
+#endif
+
 	ret = usb_autopm_get_interface(dev->intf);
 	if (ret < 0)
 		return ret;
@@ -1829,6 +2041,15 @@ static void lan78xx_get_pause(struct net
 	struct phy_device *phydev = net->phydev;
 	struct ethtool_link_ksettings ecmd;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_sw *sw = get_sw(dev);
+
+	if (sw_is_switch(sw)) {
+		phy_ethtool_ksettings_get(phydev, &ecmd);
+		return;
+	}
+#endif
+
 	phy_ethtool_ksettings_get(phydev, &ecmd);
 
 	pause->autoneg = dev->fc_autoneg;
@@ -1848,6 +2069,15 @@ static int lan78xx_set_pause(struct net_
 	struct ethtool_link_ksettings ecmd;
 	int ret;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_sw *sw = get_sw(dev);
+
+	if (sw_is_switch(sw)) {
+		phy_ethtool_ksettings_get(phydev, &ecmd);
+		return 0;
+	}
+#endif
+
 	phy_ethtool_ksettings_get(phydev, &ecmd);
 
 	if (pause->autoneg && !ecmd.base.autoneg) {
@@ -1938,6 +2168,30 @@ static const struct ethtool_ops lan78xx_
 	.get_regs	= lan78xx_get_regs,
 };
 
+#ifdef CONFIG_KSZ_SWITCH
+static int lan78xx_ioctl(struct net_device *netdev, struct ifreq *rq, int cmd)
+{
+#ifdef CONFIG_KSZ_SWITCH
+	int result;
+
+	result = sw_ioctl(netdev, rq, cmd);
+	if (result != -EOPNOTSUPP)
+		return result;
+#endif
+
+	return phy_do_ioctl_running(netdev, rq, cmd);
+}
+
+static int lan78xx_siocdevprivate(struct net_device *dev, struct ifreq *ifr,
+				  void __user *data, int cmd)
+{
+	int result;
+
+	result = sw_private_ioctl(dev, ifr, NULL, cmd);
+	return result;
+}
+#endif
+
 static void lan78xx_init_mac_address(struct lan78xx_net *dev)
 {
 	u32 addr_lo, addr_hi;
@@ -2086,8 +2340,12 @@ static int lan78xx_mdio_init(struct lan7
 		dev->mdiobus->phy_mask = ~(1 << 1);
 		break;
 	case ID_REV_CHIP_ID_7801_:
+#ifdef CONFIG_KSZ_SWITCH
+		dev->mdiobus->phy_mask = ~((1 << 1) - 1);
+#else
 		/* scan thru PHYAD[2..0] */
 		dev->mdiobus->phy_mask = ~(0xFF);
+#endif
 		break;
 	}
 
@@ -2116,6 +2374,12 @@ static void lan78xx_link_status_change(s
 {
 	struct phy_device *phydev = net->phydev;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct lan78xx_net *dev = netdev_priv(net);
+
+	if (phydev->link)
+		dev->sw_mac.port.ready = true;
+#endif
 	phy_print_status(phydev);
 }
 
@@ -2287,9 +2551,16 @@ static struct phy_device *lan7801_phy_in
 		.link = 1,
 		.speed = SPEED_1000,
 		.duplex = DUPLEX_FULL,
+		.pause = 1,
 	};
 	struct phy_device *phydev;
 
+	if (dev->chipid == ID_REV_CHIP_ID_7801_) {
+		dev->interface = PHY_INTERFACE_MODE_RGMII;
+		phydev = fixed_phy_register(PHY_POLL, &fphy_status, NULL);
+		if (phydev)
+			return phydev;
+	}
 	phydev = phy_find_first(dev->mdiobus);
 	if (!phydev) {
 		netdev_dbg(dev->net, "PHY Not Found!! Registering Fixed PHY\n");
@@ -2524,6 +2795,9 @@ static int lan78xx_change_mtu(struct net
 	if (ret < 0)
 		return ret;
 
+#ifdef CONFIG_KSZ_SWITCH
+	max_frame_len += sw_extra_mtu;
+#endif
 	ret = lan78xx_set_rx_max_frame_length(dev, max_frame_len);
 	if (!ret)
 		netdev->mtu = new_mtu;
@@ -2547,6 +2821,11 @@ static int lan78xx_set_mac_addr(struct n
 
 	eth_hw_addr_set(netdev, addr->sa_data);
 
+#ifdef CONFIG_KSZ_SWITCH
+	if (sw_mac_set_addr(netdev, &dev->sw_mac))
+		return 0;
+#endif
+
 	addr_lo = netdev->dev_addr[0] |
 		  netdev->dev_addr[1] << 8 |
 		  netdev->dev_addr[2] << 16 |
@@ -2572,6 +2851,11 @@ static int lan78xx_set_features(struct n
 	struct lan78xx_priv *pdata = (struct lan78xx_priv *)(dev->data[0]);
 	unsigned long flags;
 
+#ifdef CONFIG_KSZ_SWITCH
+	if (is_virt_dev(dev))
+		return 0;
+#endif
+
 	spin_lock_irqsave(&pdata->rfe_ctl_lock, flags);
 
 	if (features & NETIF_F_RXCSUM) {
@@ -2617,6 +2901,15 @@ static int lan78xx_vlan_rx_add_vid(struc
 	u16 vid_bit_index;
 	u16 vid_dword_index;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_sw *sw = get_sw(dev);
+
+	if (is_virt_dev(dev))
+		return 0;
+	if (sw_is_switch(sw))
+		sw->net_ops->add_vid(sw, vid);
+#endif
+
 	vid_dword_index = (vid >> 5) & 0x7F;
 	vid_bit_index = vid & 0x1F;
 
@@ -2636,6 +2929,18 @@ static int lan78xx_vlan_rx_kill_vid(stru
 	u16 vid_bit_index;
 	u16 vid_dword_index;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_sw *sw = get_sw(dev);
+
+	if (is_virt_dev(dev))
+		return 0;
+	if (sw_is_switch(sw))
+		sw->net_ops->kill_vid(sw, vid);
+#endif
+	/* Driver is being removed. */
+	if (!pdata)
+		return 0;
+
 	vid_dword_index = (vid >> 5) & 0x7F;
 	vid_bit_index = vid & 0x1F;
 
@@ -2949,6 +3254,24 @@ static int lan78xx_reset(struct lan78xx_
 	if (ret < 0)
 		return ret;
 
+	if (dev->chipid == ID_REV_CHIP_ID_7801_) {
+		ret = lan78xx_write_reg(dev, MAC_RGMII_ID,
+					MAC_RGMII_ID_TXC_DELAY_EN_);
+		if (ret < 0)
+			return ret;
+		ret = lan78xx_write_reg(dev, RGMII_TX_BYP_DLL, 0x3D00);
+		if (ret < 0)
+			return ret;
+		ret = lan78xx_read_reg(dev, HW_CFG, &buf);
+		if (ret < 0)
+			return ret;
+		buf |= HW_CFG_CLK125_EN_;
+		buf |= HW_CFG_REFCLK25_EN_;
+		ret = lan78xx_write_reg(dev, HW_CFG, buf);
+		if (ret < 0)
+			return ret;
+	}
+
 	ret = lan78xx_read_reg(dev, USB_CFG0, &buf);
 	if (ret < 0)
 		return ret;
@@ -3002,6 +3325,10 @@ static int lan78xx_reset(struct lan78xx_
 
 	lan78xx_set_multicast(dev->net);
 
+#ifdef CONFIG_KSZ_SWITCH
+	/* This resets the KSZ9897 switch. */
+#endif
+
 	/* reset PHY */
 	ret = lan78xx_read_reg(dev, PMT_CTL, &buf);
 	if (ret < 0)
@@ -3032,8 +3359,10 @@ static int lan78xx_reset(struct lan78xx_
 		return ret;
 
 	/* LAN7801 only has RGMII mode */
-	if (dev->chipid == ID_REV_CHIP_ID_7801_)
+	if (dev->chipid == ID_REV_CHIP_ID_7801_) {
 		buf &= ~MAC_CR_GMII_EN_;
+		buf |= MAC_CR_AUTO_DUPLEX_ | MAC_CR_AUTO_SPEED_;
+	}
 
 	if (dev->chipid == ID_REV_CHIP_ID_7800_) {
 		ret = lan78xx_read_raw_eeprom(dev, 0, 1, &sig);
@@ -3047,8 +3376,17 @@ static int lan78xx_reset(struct lan78xx_
 	if (ret < 0)
 		return ret;
 
+#ifdef CONFIG_KSZ_SWITCH
+	do {
+		int mtu = RX_MAX_FRAME_LEN(dev->net->mtu);
+
+		mtu += sw_extra_mtu;
+		ret = lan78xx_set_rx_max_frame_length(dev, mtu);
+	} while (0);
+#else
 	ret = lan78xx_set_rx_max_frame_length(dev,
 					      RX_MAX_FRAME_LEN(dev->net->mtu));
+#endif
 
 	return ret;
 }
@@ -3084,6 +3422,19 @@ static int lan78xx_open(struct net_devic
 	struct lan78xx_net *dev = netdev_priv(net);
 	int ret;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_mac *priv = &dev->sw_mac;
+	struct ksz_sw *sw = priv->port.sw;
+	int rx_mode = 0;
+
+	if (sw_mac_open_first(net, priv, &rx_mode)) {
+		netif_start_queue(net);
+		ret = 0;
+		goto skip_hw;
+	}
+	dev = get_hw_dev(dev);
+#endif
+
 	netif_dbg(dev, ifup, dev->net, "open device");
 
 	ret = usb_autopm_get_interface(dev->intf);
@@ -3126,10 +3477,27 @@ static int lan78xx_open(struct net_devic
 
 	netif_start_queue(net);
 
+#ifdef CONFIG_KSZ_SWITCH
+	if (sw_is_switch(sw)) {
+		sw_mac_open_next(sw, priv->hw_priv, rx_mode);
+
+skip_hw:
+		if (sw_mac_open_final(sw, net, priv->hw_priv, priv))
+			return 0;
+	}
+#endif
+
 	dev->link_on = false;
 
 	napi_enable(&dev->napi);
 
+#ifdef CONFIG_KSZ_SWITCH
+#ifdef CONFIG_KSZ_IBA_ONLY
+	if (!sw_is_switch(sw))
+		create_sw_dev(net, &dev->sw_mac);
+#endif
+#endif
+
 	lan78xx_defer_kevent(dev, EVENT_LINK_RESET);
 done:
 	mutex_unlock(&dev->dev_mutex);
@@ -3179,6 +3547,31 @@ static int lan78xx_stop(struct net_devic
 {
 	struct lan78xx_net *dev = netdev_priv(net);
 
+#ifdef CONFIG_KSZ_SWITCH
+	int iba = 0;
+
+#ifdef CONFIG_KSZ_IBA
+	iba = IBA_USE_CODE_HARD_RESET;
+#endif
+	if (sw_mac_close(net, &dev->sw_mac, iba)) {
+		bool stop_queue = true;
+
+		/* Do not shut off queue for main device. */
+#ifdef CONFIG_KSZ_IBA_ONLY
+		do {
+			struct ksz_mac *hw_priv = dev->sw_mac.hw_priv;
+
+			if (hw_priv->net == net)
+				stop_queue = false;
+		} while (0);
+#endif
+		if (stop_queue)
+			netif_stop_queue(net);
+		return 0;
+	}
+	dev = get_hw_dev(dev);
+#endif
+
 	netif_dbg(dev, ifup, dev->net, "stop device");
 
 	mutex_lock(&dev->dev_mutex);
@@ -3255,8 +3648,16 @@ static void tx_complete(struct urb *urb)
 	struct lan78xx_net *dev = entry->dev;
 
 	if (urb->status == 0) {
+#ifdef CONFIG_KSZ_SWITCH
+		struct net_device *net = skb->dev;
+
+		/* Use original network device. */
+		net->stats.tx_packets += entry->num_of_packet;
+		net->stats.tx_bytes += entry->length;
+#else
 		dev->net->stats.tx_packets += entry->num_of_packet;
 		dev->net->stats.tx_bytes += entry->length;
+#endif
 	} else {
 		dev->net->stats.tx_errors += entry->num_of_packet;
 
@@ -3277,6 +3678,14 @@ static void tx_complete(struct urb *urb)
 		case -ETIME:
 		case -EILSEQ:
 			netif_stop_queue(dev->net);
+
+#ifdef CONFIG_KSZ_SWITCH
+			do {
+				struct ksz_sw *sw = get_sw(dev);
+
+				stop_dev_queues(sw, dev->net);
+			} while (0);
+#endif
 			netif_dbg(dev, tx_err, dev->net,
 				  "tx err queue stopped %d\n",
 				  entry->urb->status);
@@ -3375,6 +3784,23 @@ lan78xx_start_xmit(struct sk_buff *skb,
 	struct lan78xx_net *dev = netdev_priv(net);
 	unsigned int tx_pend_data_len;
 
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_mac *priv = get_ksz_mac(dev);
+	unsigned long flags;
+
+	/* May be called from switch driver. */
+	if (netif_queue_stopped(net))
+		return NETDEV_TX_BUSY;
+	skb = sw_mac_tx_pre(skb, priv, 1);
+	skb = sw_mac_tx(net, skb, priv);
+	if (!skb) {
+		return NETDEV_TX_OK;
+	}
+
+	/* dev may change. */
+	dev = priv->hw_priv->dev;
+	spin_lock_irqsave(&priv->tx_lock, flags);
+#endif
 	if (test_bit(EVENT_DEV_ASLEEP, &dev->flags))
 		schedule_delayed_work(&dev->wq, 0);
 
@@ -3393,6 +3819,14 @@ lan78xx_start_xmit(struct sk_buff *skb,
 	if (tx_pend_data_len > lan78xx_tx_urb_space(dev)) {
 		netif_stop_queue(net);
 
+#ifdef CONFIG_KSZ_SWITCH
+		if (netif_queue_stopped(net)) {
+			struct ksz_sw *sw = get_sw(dev);
+
+			stop_dev_queues(sw, net);
+		}
+#endif
+
 		netif_dbg(dev, hw, dev->net, "tx data len: %u, urb space %u",
 			  tx_pend_data_len, lan78xx_tx_urb_space(dev));
 
@@ -3402,6 +3836,10 @@ lan78xx_start_xmit(struct sk_buff *skb,
 			napi_schedule(&dev->napi);
 	}
 
+#ifdef CONFIG_KSZ_SWITCH
+	spin_unlock_irqrestore(&priv->tx_lock, flags);
+#endif
+
 	return NETDEV_TX_OK;
 }
 
@@ -3439,8 +3877,12 @@ static int lan78xx_bind(struct lan78xx_n
 	if (DEFAULT_RX_CSUM_ENABLE)
 		dev->net->features |= NETIF_F_RXCSUM;
 
+#ifdef CONFIG_KSZ_SWITCH
+	dev->net->features |= NETIF_F_SG;
+#else
 	if (DEFAULT_TSO_CSUM_ENABLE)
 		dev->net->features |= NETIF_F_TSO | NETIF_F_TSO6 | NETIF_F_SG;
+#endif
 
 	if (DEFAULT_VLAN_RX_OFFLOAD)
 		dev->net->features |= NETIF_F_HW_VLAN_CTAG_RX;
@@ -3509,6 +3951,15 @@ static void lan78xx_rx_csum_offload(stru
 				    struct sk_buff *skb,
 				    u32 rx_cmd_a, u32 rx_cmd_b)
 {
+#ifdef CONFIG_KSZ_SWITCH
+	struct ksz_sw *sw = get_sw(dev);
+
+	if (sw_is_switch(sw) && using_tail_tag(sw)) {
+		skb->ip_summed = CHECKSUM_NONE;
+		return;
+	}
+#endif
+
 	/* HW Checksum offload appears to be flawed if used when not stripping
 	 * VLAN headers. Drop back to S/W checksums under these conditions.
 	 */
@@ -3535,6 +3986,29 @@ static void lan78xx_rx_vlan_offload(stru
 
 static void lan78xx_skb_return(struct lan78xx_net *dev, struct sk_buff *skb)
 {
+#ifdef CONFIG_KSZ_SWITCH
+	do {
+		struct ksz_sw *sw = get_sw(dev);
+
+		if (sw_is_switch(sw)) {
+			struct lan78xx_net *priv;
+			int rxlen = 0;
+
+			priv = sw_rx_proc(sw, skb, &rxlen);
+			if (!priv) {
+				if (rxlen) {
+					dev->net->stats.rx_packets++;
+					dev->net->stats.rx_bytes += rxlen;
+				}
+				return;
+			}
+
+			/* Use private structure in network device. */
+			dev = priv;
+		}
+	} while (0);
+#endif
+
 	dev->net->stats.rx_packets++;
 	dev->net->stats.rx_bytes += skb->len;
 
@@ -3719,7 +4193,11 @@ static int rx_submit(struct lan78xx_net
 	spin_lock_irqsave(&dev->rxq.lock, lockflags);
 
 	if (netif_device_present(dev->net) &&
+#ifdef CONFIG_KSZ_SWITCH
+	    (netif_running(dev->net) || dev->sw_mac.opened) &&
+#else
 	    netif_running(dev->net) &&
+#endif
 	    !test_bit(EVENT_RX_HALT, &dev->flags) &&
 	    !test_bit(EVENT_DEV_ASLEEP, &dev->flags)) {
 		ret = usb_submit_urb(urb, flags);
@@ -3884,6 +4362,15 @@ static void lan78xx_tx_bh(struct lan78xx
 	 */
 	netif_tx_lock(dev->net);
 	if (netif_queue_stopped(dev->net)) {
+
+#ifdef CONFIG_KSZ_SWITCH
+		if (lan78xx_tx_pend_data_len(dev) < lan78xx_tx_urb_space(dev) &&
+		    netif_queue_stopped(dev->net)) {
+			struct ksz_sw *sw = get_sw(dev);
+
+			wake_dev_queues(sw, dev->net);
+		}
+#endif
 		if (lan78xx_tx_pend_data_len(dev) < lan78xx_tx_urb_space(dev))
 			netif_wake_queue(dev->net);
 	}
@@ -4024,7 +4511,12 @@ static int lan78xx_bh(struct lan78xx_net
 	skb_queue_splice(&done, &dev->rxq_done);
 	spin_unlock_irqrestore(&dev->rxq_done.lock, flags);
 
+#ifdef CONFIG_KSZ_SWITCH
+	if (netif_device_present(dev->net) &&
+	    (netif_running(dev->net) || dev->sw_mac.opened)) {
+#else
 	if (netif_device_present(dev->net) && netif_running(dev->net)) {
+#endif
 		/* reset update timer delta */
 		if (timer_pending(&dev->stat_monitor) && (dev->delta != 1)) {
 			dev->delta = 1;
@@ -4077,6 +4569,15 @@ static int lan78xx_poll(struct napi_stru
 			} else {
 				netif_tx_lock(dev->net);
 				if (netif_queue_stopped(dev->net)) {
+
+#ifdef CONFIG_KSZ_SWITCH
+					do {
+						struct ksz_sw *sw =
+							get_sw(dev);
+
+						wake_dev_queues(sw, dev->net);
+					} while (0);
+#endif
 					netif_wake_queue(dev->net);
 					napi_schedule(napi);
 				}
@@ -4115,6 +4616,15 @@ static void lan78xx_delayedwork(struct w
 					   status);
 		} else {
 			clear_bit(EVENT_TX_HALT, &dev->flags);
+
+#ifdef CONFIG_KSZ_SWITCH
+			if (status != -ESHUTDOWN &&
+			    netif_queue_stopped(dev->net)) {
+				struct ksz_sw *sw = get_sw(dev);
+
+				wake_dev_queues(sw, dev->net);
+			}
+#endif
 			if (status != -ESHUTDOWN)
 				netif_wake_queue(dev->net);
 		}
@@ -4188,7 +4698,11 @@ static void intr_complete(struct urb *ur
 	}
 
 	if (!netif_device_present(dev->net) ||
+#ifdef CONFIG_KSZ_SWITCH
+	    (!netif_running(dev->net) && !dev->sw_mac.opened)) {
+#else
 	    !netif_running(dev->net)) {
+#endif
 		netdev_warn(dev->net, "not submitting new status URB");
 		return;
 	}
@@ -4226,6 +4740,10 @@ static void lan78xx_disconnect(struct us
 
 	netif_napi_del(&dev->napi);
 
+#ifdef CONFIG_KSZ_SWITCH
+	sw_mac_remove(dev->net, &dev->sw_mac);
+#endif
+
 	udev = interface_to_usbdev(intf);
 	net = dev->net;
 
@@ -4263,6 +4781,12 @@ static void lan78xx_tx_timeout(struct ne
 {
 	struct lan78xx_net *dev = netdev_priv(net);
 
+#ifdef CONFIG_KSZ_SWITCH
+	/* Something is wrong to reach this condition. */
+	if (is_virt_dev(dev)) {
+		return;
+	}
+#endif
 	unlink_urbs(dev, &dev->txq);
 	napi_schedule(&dev->napi);
 }
@@ -4290,7 +4814,12 @@ static const struct net_device_ops lan78
 	.ndo_change_mtu		= lan78xx_change_mtu,
 	.ndo_set_mac_address	= lan78xx_set_mac_addr,
 	.ndo_validate_addr	= eth_validate_addr,
-	.ndo_eth_ioctl		= phy_do_ioctl_running,
+#ifdef CONFIG_KSZ_SWITCH
+	.ndo_do_ioctl		= lan78xx_ioctl,
+	.ndo_siocdevprivate	= lan78xx_siocdevprivate,
+#else
+	.ndo_do_ioctl		= phy_do_ioctl_running,
+#endif
 	.ndo_set_rx_mode	= lan78xx_set_multicast,
 	.ndo_set_features	= lan78xx_set_features,
 	.ndo_vlan_rx_add_vid	= lan78xx_vlan_rx_add_vid,
@@ -4305,6 +4834,132 @@ static void lan78xx_stat_monitor(struct
 	lan78xx_defer_kevent(dev, EVENT_STAT_UPDATE);
 }
 
+#ifdef CONFIG_KSZ_SWITCH
+static int sw_mac_init(struct net_device *net, struct ksz_mac *priv)
+{
+	struct ksz_sw *sw;
+	int err;
+	int i;
+	int port_count;
+	int dev_count;
+	int mib_port_count;
+	char dev_label[IFNAMSIZ];
+	struct lan78xx_net *dev, *hw_dev;
+	struct ksz_mac *hw_priv;
+	struct net_device *main_net;
+	struct usb_interface *intf;
+	netdev_features_t features;
+
+	hw_dev = priv->dev;
+	sw = priv->port.sw;
+
+	/* This is the main private structure holding hardware information. */
+	hw_priv = priv;
+	hw_priv->net = net;
+	hw_priv->parent = sw->dev;
+	main_net = net;
+	intf = hw_dev->intf;
+
+	prep_sw_first(sw, &port_count, &mib_port_count, &dev_count, dev_label,
+		      NULL);
+	sw->msg_enable = hw_dev->msg_enable;
+
+	/* Hardware checksum cannot handle tail tag. */
+	if (dev_count > 1) {
+		main_net->features &= ~(NETIF_F_HW_CSUM | NETIF_F_RXCSUM);
+		main_net->hw_features = main_net->features;
+	}
+
+	features = main_net->features;
+
+	/* Save the base device name. */
+	strlcpy(dev_label, hw_priv->net->name, IFNAMSIZ);
+
+	spin_lock_init(&priv->tx_lock);
+	prep_sw_dev(sw, priv, 0, port_count, mib_port_count, dev_label,
+		    PHY_INTERFACE_MODE_RGMII_TXID);
+
+	INIT_DELAYED_WORK(&hw_priv->promisc_reset, promisc_reset_work);
+
+#ifdef CONFIG_KSZ_IBA_ONLY
+	INIT_WORK(&hw_priv->rmv_dev, rmv_dev_work);
+#endif
+
+	for (i = 1; i < dev_count; i++) {
+		net = alloc_etherdev(sizeof(struct lan78xx_net));
+		if (!net)
+			break;
+
+		dev = netdev_priv(net);
+		dev->intf = intf;
+		dev->net = net;
+		dev->msg_enable = hw_dev->msg_enable;
+
+		priv = &dev->sw_mac;
+		priv->hw_priv = hw_priv;
+		priv->dev = dev;
+		priv->net = net;
+
+		net->phydev = &priv->dummy_phy;
+		net->phydev->duplex = 1;
+		net->phydev->speed = SPEED_1000;
+		net->phydev->pause = 1;
+		net->phydev->autoneg = 1;
+		net->phydev->is_internal = true;
+
+		spin_lock_init(&priv->tx_lock);
+
+		net->netdev_ops = &lan78xx_netdev_ops;
+		net->watchdog_timeo = TX_TIMEOUT_JIFFIES;
+		net->ethtool_ops = &lan78xx_ethtool_ops;
+		dev_addr_set(net, main_net->dev_addr);
+
+		net->hard_header_len = main_net->hard_header_len;
+		net->hw_features = main_net->hw_features;
+		net->features = features;
+
+		SET_NETDEV_DEV(net, &intf->dev);
+
+		prep_sw_dev(sw, priv, i, port_count, mib_port_count, dev_label,
+			    PHY_INTERFACE_MODE_RGMII_TXID);
+		if (net->phydev->mdio.bus)
+			net->phydev->adjust_link = sw_adjust_link;
+		else
+			net->phydev->mdio.addr = priv->phy_addr;
+
+		err = register_netdev(net);
+		if (err) {
+			free_netdev(net);
+			break;
+		}
+
+		netif_carrier_off(net);
+	}
+
+#ifndef CONFIG_KSZ_IBA_ONLY
+	/*
+	 * Adding sysfs support is optional for network device.  It is more
+	 * convenient to locate eth0 more or less than spi<bus>.<select>,
+	 * especially when the bus number is auto assigned which results in a
+	 * very big number.
+	 */
+	err = init_sw_sysfs(sw, &hw_priv->sysfs, &main_net->dev);
+
+#ifdef CONFIG_1588_PTP
+	if (sw->features & PTP_HW)
+		err = init_ptp_sysfs(&hw_priv->ptp_sysfs, &main_net->dev);
+#endif
+#ifdef CONFIG_KSZ_DLR
+	if (sw->features & DLR_HW)
+		err = init_dlr_sysfs(&main_net->dev);
+#endif
+	sw_device_seen++;
+#endif
+
+	return 0;
+}
+#endif
+
 static int lan78xx_probe(struct usb_interface *intf,
 			 const struct usb_device_id *id)
 {
@@ -4438,6 +5093,17 @@ static int lan78xx_probe(struct usb_inte
 	/* driver requires remote-wakeup capability during autosuspend. */
 	intf->needs_remote_wakeup = 1;
 
+#ifdef CONFIG_KSZ_SWITCH
+	/* Point to real private structure holding hardware information. */
+	setup_ksz_mac(dev, netdev);
+
+	sw_mac_init_pre();
+
+#ifndef CONFIG_KSZ_IBA_ONLY
+	ret = sw_mac_chk(&dev->sw_mac);
+#endif
+#endif
+
 	ret = lan78xx_phy_init(dev);
 	if (ret < 0)
 		goto out7;
@@ -4458,6 +5124,15 @@ static int lan78xx_probe(struct usb_inte
 	pm_runtime_set_autosuspend_delay(&udev->dev,
 					 DEFAULT_AUTOSUSPEND_DELAY);
 
+#ifdef CONFIG_KSZ_SWITCH
+	dev->sw_mac.saved_phy = netdev->phydev;
+
+#ifndef CONFIG_KSZ_IBA_ONLY
+	if (dev->sw_mac.port.sw)
+		sw_mac_init(netdev, &dev->sw_mac);
+#endif
+#endif
+
 	return 0;
 
 out8:
@@ -4905,6 +5580,14 @@ static bool lan78xx_submit_deferred_urbs
 		} else {
 			if (ret == -EPIPE) {
 				netif_stop_queue(dev->net);
+
+#ifdef CONFIG_KSZ_SWITCH
+				do {
+					struct ksz_sw *sw = get_sw(dev);
+
+					stop_dev_queues(sw, dev->net);
+				} while (0);
+#endif
 				pipe_halted = true;
 			} else if (ret == -ENODEV) {
 				netif_device_detach(dev->net);
