--- ptp_clock.c	2019-10-10 13:54:51.021668452 -0700
+++ ptp_clock.c.new	2019-10-07 19:01:42.615035847 -0700
@@ -157,6 +157,17 @@
 		ptp->dialed_frequency = tx->freq;
 	} else if (tx->modes == 0) {
 		tx->freq = ptp->dialed_frequency;
+#if defined(CONFIG_KSZ_PTP) || defined(CONFIG_LAN937X_PTP)
+		do {
+			s64 freq;
+
+			err = ops->gettime64(ops, NULL);
+			freq = err;
+			freq <<= 13;
+			freq = div64_s64(freq, 125);
+			tx->freq = (s32) freq;
+		} while (0);
+#endif
 		err = 0;
 	}
 
