diff -Napur linux-at91-linux4sam-2021.04-orig/net/bridge/br_device.c linux-at91-linux4sam-2021.04/net/bridge/br_device.c
--- linux-at91-linux4sam-2021.04-orig/net/bridge/br_device.c	2021-07-29 07:58:03.000000000 -0700
+++ linux-at91-linux4sam-2021.04/net/bridge/br_device.c	2021-09-01 14:23:52.943566095 -0700
@@ -456,6 +456,9 @@ void br_dev_setup(struct net_device *dev
 	dev->hw_features = COMMON_FEATURES | NETIF_F_HW_VLAN_CTAG_TX |
 			   NETIF_F_HW_VLAN_STAG_TX;
 	dev->vlan_features = COMMON_FEATURES;
+#ifdef CONFIG_KSZ_SWITCH
+	dev->hard_header_len += 6;
+#endif
 
 	br->dev = dev;
 	spin_lock_init(&br->lock);
