diff -Napur linux-at91-linux4sam-2020.10-orig/drivers/pinctrl/pinctrl-at91.c linux-at91-linux4sam-2020.10/drivers/pinctrl/pinctrl-at91.c
--- linux-at91-linux4sam-2020.10-orig/drivers/pinctrl/pinctrl-at91.c	2021-01-14 04:54:56.000000000 -0800
+++ linux-at91-linux4sam-2020.10/drivers/pinctrl/pinctrl-at91.c	2021-08-27 14:49:32.488003839 -0700
@@ -1541,8 +1541,13 @@ static void gpio_irq_unmask(struct irq_d
 	struct at91_gpio_chip *at91_gpio = irq_data_get_irq_chip_data(d);
 	void __iomem	*pio = at91_gpio->regbase;
 	unsigned	mask = 1 << d->hwirq;
+#if 1
+	unsigned long	isr;
 
 	if (pio)
+		isr = readl_relaxed(pio + PIO_ISR);
+#endif
+	if (pio)
 		writel_relaxed(mask, pio + PIO_IER);
 }
 
