--- br_device.c.orig	2019-10-29 01:20:09.000000000 -0700
+++ br_device.c	2020-01-24 18:08:01.881885626 -0800
@@ -421,6 +421,9 @@ void br_dev_setup(struct net_device *dev
 	dev->hw_features = COMMON_FEATURES | NETIF_F_HW_VLAN_CTAG_TX |
 			   NETIF_F_HW_VLAN_STAG_TX;
 	dev->vlan_features = COMMON_FEATURES;
+#ifdef CONFIG_KSZ_SWITCH
+	dev->hard_header_len += 6;
+#endif
 
 	br->dev = dev;
 	spin_lock_init(&br->lock);
