diff -Napur linux-5.10.68-orig/drivers/mfd/at91-usart.c linux-5.10.68/drivers/mfd/at91-usart.c
--- linux-5.10.68-orig/drivers/mfd/at91-usart.c	2021-09-22 03:28:07.000000000 -0700
+++ linux-5.10.68/drivers/mfd/at91-usart.c	2023-03-15 19:37:23.372489570 -0700
@@ -15,15 +15,11 @@
 #include <linux/of.h>
 #include <linux/property.h>
 
-static const struct mfd_cell at91_usart_spi_subdev = {
-	.name = "at91_usart_spi",
-	.of_compatible = "microchip,at91sam9g45-usart-spi",
-};
+static const struct mfd_cell at91_usart_spi_subdev =
+	MFD_CELL_NAME("at91_usart_spi");
 
-static const struct mfd_cell at91_usart_serial_subdev = {
-	.name = "atmel_usart_serial",
-	.of_compatible = "atmel,at91rm9200-usart-serial",
-};
+static const struct mfd_cell at91_usart_serial_subdev =
+	MFD_CELL_NAME("atmel_usart_serial");
 
 static int at91_usart_mode_probe(struct platform_device *pdev)
 {
